pymeasure/instruments/hp/hp3478A.py

```diff
@@ -0,0 +1,491 @@
+        ("cal_enable", c_uint8, 1),
+        ("ext_trig",   c_uint8, 1),
+        ("res1",       c_uint8, 1),
+
+        #Byte 3: Serial Poll Mask (SRQ)
+        ("SRQ_data_rdy",         c_uint8, 1),
+        ("res2",                 c_uint8, 1),
+        ("SRQ_syntax_error",     c_uint8, 1),
+        ("SRQ_internal_error",   c_uint8, 1),
+        ("SRQ_front_panel",      c_uint8, 1),
+        ("SRQ_cal_error",        c_uint8, 1),
+        ("res3",                 c_uint8, 1),
+        ("SRQ_PON",              c_uint8, 1),
+
+        #Byte 4: Error Information
+        ("ERR_cal",        c_uint8, 1),
+        ("ERR_RAM",        c_uint8, 1),
+        ("ERR_ROM",        c_uint8, 1),
+        ("ERR_slope",      c_uint8, 1),
+        ("ERR_AD",         c_uint8, 1),
+        ("ERR_AD_Link",    c_uint8, 1),
+        ("res4",           c_uint8, 1),
+        ("res5",           c_uint8, 1),
+
+        #Byte 5: DAC Value
+        ("DAC_value",       c_uint8, 8),
+    ]
+    def __str__(self):
+        """
+        Returns a pretty formatted (human readable) string showing the status of the instrument
+
+        """
+        inv_modes = dict(zip(HP3478A.MODES.values(),HP3478A.MODES.keys()))
+        cur_mode = inv_modes["F" + str(self.function)]
+        cur_range = list(HP3478A.RANGES[cur_mode].keys())[self.range - 1]
+        if cur_range >= 1E6:
+            r_str = str(cur_range / 1E6) + ' M'
+        elif cur_range >= 1000:
+            r_str = str(cur_range / 1000) + ' k'
+        elif cur_range <= 1:
+            r_str=str(cur_range * 1000) + ' m'
+        else:
+            r_str=str(cur_range) + ' '
+        return (
+            "function: {}, range: {}, digits: {}\
+                \nStatus:\n  internal | external trigger: {} | {}\n  Auto ranging: {}\n  AutoZero: {}\
+                \n  50Hz mode: {}\n  Front/Rear selection: {}\n  Calibration enable: {}\
+                \nSerial poll mask (SRQ):\n  SRQ for Data ready: {}\
+                \n  SRQ for Syntax error: {}\n  SRQ for Internal error: {}\n  SRQ Front Panel button: {}\
+                \n  SRQ for Cal err: {}\n  SQR for Power on: {}\
+                \nError information: \n Calibration: {}  RAM: {}  ROM: {}  AD Slope: {} AD: {} AD-Link: {} \
+                \nDAC value: {}".format(
+                cur_mode, r_str, 6-self.digits, self.int_trig, self.ext_trig, self.auto_range,
+                self.auto_zero, self.fifty_hz, self.front_rear , self.cal_enable,
+                self.SRQ_data_rdy, self.SRQ_syntax_error, self.SRQ_internal_error, self.SRQ_front_panel, self.SRQ_cal_error, self.SRQ_PON,
+                self.ERR_cal, self.ERR_RAM, self.ERR_ROM, self.ERR_slope, self.ERR_AD, self.ERR_AD_Link,
+                self.DAC_value)
+            )
+
+class Status(ctypes.Union):
+    """Union type element for the decoding of the status bit-fields"""
+    _fields_ = [
+        ("B", Status_bytes),
+        ("b", Status_bits)
+    ]
+
+# @staticmethod does not work ?!
```

Comment:
That's a bit strange, I would expect this would work just fine as a staticmethod; can you tell me what happens?

Reply:
with 

```
@staticmethod

def get_status(status_bytes, field=None):
    ...

```
I get the following error:

```

Traceback (most recent call last):

  File "/home/robby/RF_stuff/rf-detectors/soft/xxx/xxx/test.py", line 37, in <module>
    print(str(meter.status))

  File "/home/robby/builds/pymeasure/pymeasure/instruments/instrument.py", line 220, in fget
    value = get_process(vals[0])

  File "/home/robby/builds/pymeasure/pymeasure/instruments/hp/hp3478A.py", line 426, in <lambda>
    get_process = (lambda x: get_status(x.encode(encoding="ASCII"))),

TypeError: 'staticmethod' object is not callable

```
Could this come from the fact that get_status() is defined outside the HP3478A class?

Reply:
A staticmethod is defined within a class, so I am not sure what you mean with "outside the HP3478A class" ? Where get_status is defined?

Normally you should call the staticmethod as an attribute of the class where it's defined.

Have a look at https://pythonbasics.org/static-method/

Reply:
Ahh yes, that would indeed be the problem. I think, for the sake of code clarity, it'd be nice to place these 'get' functions as staticmethods in the class; that way they are nicely confined within the class that uses them and get nicely documented.

In some of the functions you actually use the class (e.g. `HP3478A.MODES`), so these you can best add as `@classmethod` rather than as `@staticmethod` such that you can access the `MODES`:
```
@classmethod
def get_mode(cls, status_bytes):
    ....
    ....
    inv_modes = dict(zip(cls.MODES.values(),cls.MODES.keys()))
    ....
    ....
```

I'd also suggest, to prevent confusion about what these methods do, to change the `get` into something like `decode` (as you use in the docstring) (i.e. `decode_range`).

(btw: I think for inverting a dictionary, a nicer/maybe clearer way is to use `inv_modes = {v: k for k, v in MODES.items()}`

Reply:
Thanks @msmttchr & @CasperSchippers again for your input and sticking with me on my first bigger python project.

@msmttchr 
the function definition for get_status and the other get_functions is in the same file, but before the definition of the class HP3478A:
![20210717_spyder](https://user-images.githubusercontent.com/17066917/126037794-e8223ab6-d4fc-49be-a4ce-f31928013913.png)

But now I understand that the staticmethod or classmethod only work within the class


I will do the folloing:
1. change the naming of the get_status etc functions to decode_status
2. move these functions into the class hp3478a
3. improve the inverting of the dictionary